docs/worker-versioning.md

```diff
@@ -0,0 +1,229 @@
+of each Task Queue.
+
+There are two types of rules: _Build ID Assignment rules_ and _Compatible Build ID Redirect rules_.
+
+### Assingment Rules
+Assignment rules are used to assign a Build ID for a new execution when it starts. Its primary
+use case is to specify the latest Build ID, but it has powerful features for gradual rollout
+of a new Build ID.
+
+Once a Build ID is assigned to a Workflow Execution, and it completes its first Workflow Task,
+the workflow stays on the assigned Build ID regardless of changes in Assignment rules. This
+eliminates the need for compatibility between versions when you only care about using the new
+version for new Workflows and let existing Workflows finish in their own version.
+
+Activities, Child Workflows and Continue-as-New executions have the option to inherit the 
+Build ID of their parent/previous Workflow or use the latest Assignment rules to independently 
+select a Build ID. This is specified by the parent/previous Workflow using VersioningIntent.
+
+In absence of (applicable) Redirect rules the task will be dispatched to Workers 
+of the Build ID determined by the Assignment rules (or inherited). Otherwise, the final 
+Build ID will be determined by the Redirect rules.
+
+When using Worker Versioning on a Task Queue, in the steady state,
+there should typically be a single assignment rule to send all new executions
+to the latest Build ID. Existence of at least one such "unconditional"
+rule at all times is enforces by the system, unless the `force` flag is used
+by the user when replacing/deleting these rules (for exceptional cases).
+
+During a deployment, one or more additional rules can be added to assign a
+subset of the tasks to a new Build ID based on a "ramp percentage".
+
+When there are multiple assignment rules for a Task Queue, the rules are
+evaluated in order, starting from index 0. The first applicable rule will be
+applied and the rest will be ignored.
+
+In the event that no assignment rule is applicable on a task (or the Task
+Queue is simply not versioned), the tasks will be dispatched to an
+unversioned Worker.
+
+### Redirect Rules
+Redirect rules should only be used when you want to move workflows and activities assigned to
+one Build ID (source) to another compatible Build ID (target). You are responsible to make sure
+the target Build ID of a redirect rule is able to process event histories made by the source
+Build ID by using [Patching](https://docs.temporal.io/workflows#patching) or other means.
+
+Most deployments are not expected to need these rules, however following
+ situations can greatly benefit from redirects:
+  - Need to move long-running Workflow Executions from an old Build ID to a
+    newer one.
+  - Need to hotfix some broken or stuck Workflow Executions.
+
+In steady state, redirect rules are beneficial when dealing with old
+Executions ran on now-decommissioned Build IDs:
+  - To redirecting the Workflow Queries to the current (compatible) Build ID.
```

Comment:
Hmm, thanks for pointing this out.  It's concerning as it implies that we will break workflow queries on closed workflows by default unless people use redirect rules.  Which seems to not be what you're recommending above.
Thinking about target audience, if the readers are platform developers, they'll want to set up a deploy system that accounts for all the different things their developers might do, including querying of closed workflows, even though that behavior might be unusual.
If developers just own a worker with a couple workflows with few plans for growth, they might be able to get away with some shortcuts.

Reply:
They'll only break if the original workers are decommissioned. Which is certainly a tempting thing to do if those workers are no longer processing open workflows. I could imagine many situations where leaving a single old worker running to handle queries makes sense (instead of redirecting the queries).